rfc/redis-removal.md

```diff
@@ -0,0 +1,171 @@
+    decide if the timeline should be delayed
+
+Non-Goals:
+
+*   Debating if SQL is the best replacement default persistence backend to
+    standardize on
+*   Documenting and solving all technical issues for each service to migrate;
+    those will be handled separately whereas this document concerns itself
+    primarily with the larger processes and timeline
+*   Identification of all new processes for maintaining services on a SQL
+    backend; these will be handled separately
+*   Prescription of what SQL vendor to use: Current functionality has been built
+    using MySQL; support for other vendors will be addressed separately
+
+## Case Against Redis
+
+Redis was introduced into the project as an easy way to get started. It's
+phenomenally flexible, and a single, common data store simplifies the
+operational story of the large operational footprint that Spinnaker has.
+Further, it was useful in that its flexibility meant services could be rapidly
+iterated on, as no schema was needed from the start.
+
+Unfortunately, Redis is difficult to operationalize as a persistent datastore,
+which many of the services treat Redis as. Community members are often surprised
+when Redis restarts to find Spinnaker has lost in-flight work. Operating Redis
+beyond a cache is not a skill that many organizations have. Worse yet, the level
+of effort to support a highly-available, persistent story for Redis is very high
+and littered with edge cases. For example, when tested at Netflix, both Redis
+Cluster and Sentinel regularly encountered incorrect master failover behavior
+that would lead to corrupted clusters or lost data.
+
+Furthermore, developing applications on Redis is typically a foreign concept and
+offers a confusing touch point for anyone looking to get involved in
+contributing.
+
+Having a singular datastore for Spinnaker is still necessary, as it eases the
+burden of onboarding. Now that Spinnaker is no longer a young project, the need
+for fast service iteration has largely dissapated. SQL was chosen as the
+replacement because modern SQL is flexible and well understood.
+
+## Service Inventory
+
+Most of the larger services already have support for SQL.
+
+* Full support: ✔️
+* Partial support: ~
+* No support: ✗
+
+| | |
+|-|-|
+| Clouddriver | ✔️ |
+| Echo | ✗ |
+| Fiat | ✗ |
+| Front50 | ✔️ |
+| Gate | ~ |
+| Igor | ✗ |
+| Kayenta | ~ |
+| Keel | ✔️ |
+| Orca | ✔️ |
+| Rosco | ✗ |
+| Swabbie | ✗ |
+
+## Service Migration Process
+
+_Some words on how services should be written such that they can be migrated
+without downtime._
+
+Being able to migrate existing deployments from Redis to SQL for each particular
+service will be crucial. Each service must provide these features and
+characteristics before they can be said to having "full support":
+
+* A service must support a dual-backend mode, where writes are sent to the SQL
+  store while reads can still hit Redis
```

Comment:
I don't think this is going to work? If you're not writing to both places, then any reads that go to Redis are just going to serve you some random stale data... whatever was in there when you turned off writes.

Reply:
For clouddriver, orca and front50 we've been sending new writes to the new system, then reading from both the old & new. Seems to have worked fine so far. Then the background worker moves over any data not yet in the new datastore.

Reply:
By "reads can still hit Redis" do you mean all reads *will* hit Redis (as well as MySQL).  I think you're describing what `DualExecutionRepository` does in Orca where it hits both repositories for all reads and returns the superset (roughly speaking)...but the way it's written makes it seem some reads will only hit redis.

Reply:
Yes, that's what I meant. I'll make it clearer.